stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,234 @@
+#ifndef STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+#define STAN_MATH_REV_FUN_HMM_MARGINAL_LPDF_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/row.hpp>
+#include <stan/math/prim/fun/col.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
+#include <stan/math/prim/fun/exp.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <stan/math/prim/core.hpp>
+#include <Eigen/Core>
+#include <vector>
+#include <iostream>
+
+namespace stan {
+namespace math {
+
+/**
+ * For a Hidden Markov Model with observation y, hidden state x,
+ * and parameters theta, return the log marginal density, log
+ * pi(y | theta). In this setting, the hidden states are discrete
+ * and take values over the finite space {1, ..., K}.
+ * The marginal lpdf is obtained via a forward pass.
+ * The [in, out] argument are saved so that we can use them when
+ * calculating the derivatives.
+ *
+ * @param[in] log_omega log matrix of observational densities.
+ *              The (i, j)th entry corresponds to the
+ *              density of the ith observation, y_i,
+ *              given x_i = j.
+ * @param[in] Gamma transition density between hidden states.
+ *              The (i, j)th entry is the probability that x_n = j,
+ *              given x_{n - 1} = i. The rows of Gamma are simplexes.
+ * @param[in] rho initial state
+ * @param[in, out] alphas unnormalized partial marginal density.
+ *                   The jth column is the joint density over all
+ *                   observations y and the hidden state j.
+ * @param[in, out] alpha_log_norms max coefficient for column of alpha,
+ *                   to be used to normalize alphas.
+ * @param[in, out] omegas term-wise exponential of omegas.
+ * @return log marginal density.
```

Comment:
This function is for internal use. Alpha and omega are saved so that I can use when computing derivatives. Only the templated version, with partials and operands, is exposed to the user. If needed I can rename the internal function.